src/Sleep.js

```diff
@@ -0,0 +1,106 @@
+class Sleep {
+  constructor(sleepData) {
+    this.sleepData = sleepData;
+  }
+
+  getUserData(id) {
+    return this.sleepData.filter(obj => {
+      return obj.userID === id;
+    });
+  }
+
+  calculateAverageHours(id) {
+    return this.getUserData(id).reduce((acc, object) => {
+      return Math.round(acc += object.hoursSlept / this.getUserData(id).length);
+    }, 0);
+  }
+
+  calculateAverageQuality(id) {
+    return this.getUserData(id).reduce((acc, object) => {
+      return Math.round(acc += object.sleepQuality / this.getUserData(id).length);
+    }, 0);
+  }
+
+  getDayHours(id, date) {
+    return this.getUserData(id).find(day => day.date === date).hoursSlept;
+  }
+
+  getDayQuality(id, date) {
+    return this.getUserData(id).find(day => day.date === date).sleepQuality;
+  }
+
+  getHoursByWeek(id, date) {
+    let endDate = this.getUserData(id).findIndex(object => object.date === date);
+    let filteredData = this.getUserData(id).map(day => day.hoursSlept);
+    return endDate - 6 >= 0 ?
+      filteredData.slice(endDate - 6, endDate + 1) :
+      filteredData.slice(0, endDate + 1);
+  }
+
+  getQualityByWeek(id, date) {
+    let endDate = this.getUserData(id).findIndex(object => object.date === date);
+    let filteredData = this.getUserData(id).map(day => day.sleepQuality);
+    return endDate - 6 >= 0 ?
+      filteredData.slice(endDate - 6, endDate + 1) :
+      filteredData.slice(0, endDate + 1);
+  }
+
+  getAverageQualityOfUsers() {
+    const average = this.sleepData.reduce((acc, object) => {
+      return acc += object.sleepQuality / this.sleepData.length;
+    }, 0 )
+    return average.toFixed(2)
+  }
+
+  findGoodSleepers(date) {
+    let goodSleepers = [];
+    let ids = [];
+    this.sleepData.forEach(obj => {
+      if (!ids.includes(obj.userID)) {
+        ids.push(obj.userID);
+      }
+    });
+    ids.forEach(id => {
+      let endDate = this.getUserData(id).findIndex(object => object.date === date);
+      let filteredData = this.getUserData(id).map(day => day.sleepQuality);
+      let weeklyNumbers = endDate - 6 >= 0 ?
+        filteredData.slice(endDate - 6, endDate + 1) :
+        filteredData.slice(0, endDate + 1);
+      let average = weeklyNumbers.reduce((acc, num) => {
+        return (acc += num / weeklyNumbers.length);
+      }, 0)
+      if (average >= 3) {
+        goodSleepers.push(id);
+      }
+    });
+    return goodSleepers;
+  }
```

Comment:
can definitely refactor this, we already have a method to find weekly quality for a user: 
```
findGoodSleepers(date) {
  let userIDs = [...new Set(this.data.map(object => object.userID))];
  userIDs.filter(user => this.getQualityByWeek(user, date) >= 3);
}
```
The first line gets us an array of unique userIDs, the second filters through them using our other method

Reply:
Same question as above

Reply:
Open to changing as it is significantly shorter. Just need know what it means.

I get ... copies an identified array and then properties can be overwritten. I dont know what "new Set" is in regards to the ...

Reply:
Set is a class that basically gets rid of duplicates, but its an object so we convert it back to an array with the spread operator [...]

Here's some good articles about the two:
[new Set?](https://www.geeksforgeeks.org/sets-in-javascript/)
[[...]?](https://zendev.com/2018/05/09/understanding-spread-operator-in-javascript.html)

Reply:
So essentially, userIDs is equal to a new Set(which is an object with no duplicates). The spread operator takes that object and takes its valuse and turns it into an array. Our ...new Set is going over every ID  in our data set. It originally looks like
(1 ,1,1,1,1,1,11,11,11,11,11,111,111 etc.). But since we are creating a new set no duplicates occur. So our array really looks like [1, 11, 111].

After that we are filtering through our UserIDS array we just created. We then invoke our function from previous line and pass in user to said function (the ID) and then pass in the date from the top of the function and see if its >= 3.

Reply:
Yes, exactly, we want an array of userIDs where they have high quality sleep, so we get an array of all the user IDs and filter it with our method that finds sleep quality

Reply:
Perfect. Switching now

Reply:
So this doesnt work. Im trying to get it too but if I cant soon. Lets just merge and deal with it another time